infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java

```diff
@@ -332,12 +325,11 @@ private static void clearDataFiles(ServerContext serverContext) {
      */
     static void restartServer(boolean resetData) {
 
-        if (resetData) {
-            cleanupServer = true;
-        }
-
         log.info("RestartServer: Shutting down corfu server");
         activeServer.close();
+        if (resetData) {
+            clearDataFiles(activeServer.getServerContext());
```

Comment:
There's a race between this and the main thread. While data is being deleted a new context can be created at the same time.

Reply:
The previous version also had a similar problem - we delete data and close streamlog in parallel at the same time.

Fixed, by introducing and using `CorfuServerStateMachine` which brings total ordering between start, stop, reset_log_unit, shutdown operations